### DISCUSSION TITLE
Local packages should have realistic module resolution that mimics installing a package from a registry.

### DISCUSSION THREAD

**Discussion by ChocolateLoverRaj** [npm/feedback#384] [CLI] [answered]
## Introduction
It is common for a GitHub repo to have multiple related npm packages in it.

Example package structure for development
- a
  - package.json (depends on `last-element`)
  - index.js
- b
  - package.json (depends on `last-element` and `file:../some-package`)
  - index.js

Doing `npm i` in both `a` and `b` will result in two copies of `last-element`. This makes sense because both the packages depend on `last-element`.

However, the module resolution does not make sense. Let's say `a/index.js` is
```js
console.log(require.resolve('last-element'))
```
and `b/index.js` is
```js
require('a')
console.log(require.resolve('last-element'))
```

Now if we run `node index` in `a`, it will print (relative to root dir) 
```
a/node_modules/last-element/last.js`
```
This is good.

If we run `node index` in `b`, it will print (relative to root dir)
```
a/node_modules/last-element/last.js
b/node_modules/last-element/last.js
```
*\*Confusion*\*. This is not how module resolution should be. The first line should also be `b/node_modules/last-element/last.js`, because it is a dependency of `b`.

## Why
Local dependencies should mimic installing the dependency from a registry, or at least have an option to.

Let's say we published the package `a` with the version `1.0.0`, and had a package consumer with the following `package.json`:
```json
{
  "private": true,
  "name": "my-website",
  "dependencies": {
    "a": "^1.0.0"
  }
}
```

Now let's say it has the same exact `index.js` as `b/index.js` in the consumer's folder and then ran `node index`.
This would be the output, with the paths relative to the consumer's folder.
```
node_modules/last-element/last.js
node_modules/last-element/last.js
```
The to paths are the same.

## Problem with current resolution
Many times, it's okay to use two different copies of a package. In the example package `last-element`, there is no state - it just returns a value given a value. This is not the case with other packages such as [`graceful-fs`](https://www.npmjs.com/package/graceful-fs) and [`react`](https://www.npmjs.com/package/react). These two packages use global state which is referenced / updated by all files depending on it. A common error in `react` is an [invalid hook call](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react). If the example package `a` and `b` were react packages with `react` installed and used hooks, they would get the invalid hook call error in react if run from `b`.

Sample problems caused by current module resolution
- [Setup multiple local modules with npm link + dedupe](https://github.com/npm/npm/issues/7742)
- [Invalid hook call in 9.0.6](https://github.com/vercel/next.js/issues/9022)
- [How to fix NPM link duplicate dependencies issues](https://blog.maximeheckel.com/posts/duplicate-dependencies-npm-link)

### Change default behavior
In npm 8, the module resolution could be changed.

### Flag when installing
Maybe there could be a `--mimic-registry-install` flag when doing `npm i`, which changes the `package-lock.json` to tell npm to use realistic module resolution. 

### Option in `package.json`
There could be a `localResolution` option in `package.json`, which could be set to `mimicRegistryDependency` or an option with similar wording.

### Pnpm
Pnpm has [workspaces](https://pnpm.io/workspaces#shared-workspace-lockfile), which makes sure only one version of a package is installed in a workspace (I think even without workspaces, only one version of a package is installed in a computer). So we can create a ` pnpm-workspace.yaml` file, and add `"a": "workspace:^1.0.0"` as a dependency of `b`.

Result after running `node index` in `b` (relative to the folder which contains `a` and `b`).
```
node_modules\.pnpm\last-element@1.0.1\node_modules\last-element\last.js
node_modules\.pnpm\last-element@1.0.1\node_modules\last-element\last.js
```

**Comment by MylesBorins** [comment, marked as answer]
@ChocolateLoverRaj as of npm 7 we do support workspaces

https://docs.npmjs.com/cli/v7/using-npm/workspaces

Would this solve the issue you are running into?

**Reply by ChocolateLoverRaj** [reply]
No. I tried and it still had the two different files. I think using workspaces works the same as `file:` protocol when developing packages.

**Reply by MylesBorins** [reply]
If you use workspaces and two packages have the same dependency and same version it will end up in a single `node_modules` folder at the root of the workspace.

How Node.js resolves files is not something npm is in control of, although we are in control of how files are reified on disk.

If you think this is a bug you should open an issue in https://github.com/npm/cli/issues

If this is a more extensive feature request you should open an issue over at https://github.com/npm/rfcs
